### ISSUE TITLE
No connected devices found.

### ISSUE THREAD

**Issue by gabrieldavison** [mossblaser/plottie#10]
I am trying to use plottie with a silhouette cameo 4 plus on the latest firmware and when I try to plot an svg I get this error : `plottie: error: No connected devices found.` . When I run `plottie --ls` nothing is printed to the terminal. I am able to print stuff using silhouette studio so the printer is definitely working. Do you have any idea on how I might fix this / how I can debug this further? For context I am using an M1 macbook pro on MacOS 13.2. Thanks

**Comment by mossblaser** [comment]
Hey there!

It looks like you might be the first person to try that device with plottie (or, more specifically, [`py_silhouette`](https://github.com/mossblaser/py_silhouette), plottie's sister project which actually does the "talk to the plotter" bit.

You'll need to find the USB vendor ID and product ID that the device advertises over USB. The vendor ID will almost certainly be `0x0B4D` but the product ID will be new for the Cameo 4 Plus. Under Linux there's an `lsusb` tool which can straightforwardly give you this information. I imagine there's a similar (or even identical) command under MacOS -- or possibly some settings widget somewhere.

You could test out adding this to a new `SUPPORTED_DEVICE_PARAMETERS` entry into the [`device.py`](https://github.com/mossblaser/py_silhouette/blob/master/py_silhouette/device.py#L169) file in `py_silhouette`. Assuming that the 4 Plus is just a wider version of the Cameo 4 the following based on the Cameo 4 will probably be a good starting point:

```python
    DeviceParameters(
        product_name='Silhouette Cameo4 Plus',
        usb_vendor_id=0x0B4D,
        usb_product_id=0x????,  # Figure this number out!
        area_width_min=inch2mm(3.0),
        area_width_max=inch2mm(15.0),
        area_height_min=inch2mm(3.0),
        area_height_max=inch2mm(40.0),
        tool_depth_min=0,
        tool_depth_max=10,
        tool_speed_max=3000.0,
    ),
```

You can find the relevant file on your system by running the following in your Python environment:

```python
>>> import py_silhouette
>>> py_silhouette.device.__file__
```

Once you've added the new device details Plottie should be able to find it and do its thing. If you can, I'd be grateful if you could give it a spin and let me know if it seems to work for you (specifically things like setting the speed to 100% appears to make it go as fast as you'd expect! The Cameo 4 series seem to be faster than the rest of the range).

If all looks good, let me know the magic product ID number here and I'll update `py_silhouette` accordingly.

Thanks in advance for your help!

**Comment by gabrieldavison** [comment]
Thanks so much for your help, so the device ID is `0x1138`. I am now getting a response from the printer when I try to print, however I haven't managed to get it to actually print yet. The pen just quickly moves up and down a few times and then the print stops. I think this could be to do with the paths being generated incorrectly. I don't think this is related to the printer as I am getting a strange output when I use the `-D` flag to generate an svg from the included cutting_example.svg (all the paths seem to be located in the top left corner.) I have included the output file so you can see what I mean. Im wondering if there is something else going on here because as far as i can tell the `-D` output shouldn't be effected by my printer?

![test](https://github.com/mossblaser/plottie/assets/18332672/4f79a508-23f0-4c34-9f03-063ed5ad5762)

**Comment by mossblaser** [comment]
Uh-oh! You're right, this is not related to your device: well done for figuring out that something was going wrong! (It must have been a little alarming to see your machine randomly dot around without managing to do anything!)

Looks like something has changed in PySide6 (the Qt bindings for Python which are internally used by [svgoutline](https://github.com/mossblaser/svgoutline) process SVGs into line segments for Plottie) in the recently released version 6.7.0.

If you force a downgrade using `pip install "pyside6<6.7.0"` you will hopefully find everything should start working right. In the meantime I'll have to work out what's changed and how to fix it...

Assuming that this change gets you going, please do let me know if everything seems to work well for your plotter and I'll get the device info added to `py_silhouette`.

Thanks again!

**Comment by mossblaser** [comment]
Thanks again for finding the above bug, its now fixed and released and you should be able to install the fixed release using:

```
$ pip install svgoutline==0.2.1
```

Please do let me know how you get on with your plotter and if all appears to work well I'll add the device ID to `py_silhouette`.

Thanks again,

Jonathan

**Comment by gabrieldavison** [comment]
Hey sorry for the delayed reply I have been away so I didn't get a chance to test this. Im pleased to say everything is working now. Thanks so much for your help, I am very happy to not need the silhouette software anymore. I think you are right about the speed being a little bit fast with this printer but that's tuneable with the speed parameter for now. I will run some more tests when I get a chance and try to find values that match the ones from the silhouette software. Is this tuned with `tool_speed_max` in the device parameters? If so will open up a PR in py_silhouette when I have had time to test

**Comment by mossblaser** [comment]
No worries! Thanks for looking into this.

Do let me know how you get on with the speed settings and I'll set according to your experiments! Once you've got a sensible value I'll add the device!

Cheers,

Jonathan

**Comment by arthursw** [comment]
Here is the settings for the Camero Pro 4: 

```

    # Contributed by Yohalmo Garcia (@yohalmogarcia on GitHub, see issue #3)
    # and @LoveisAFK on GitHub (see issue #6)
    DeviceParameters(
        product_name='Silhouette Cameo4',
        usb_vendor_id=0x0B4D,
        usb_product_id=0x1139,
        area_width_min=inch2mm(3.0),
        area_width_max=inch2mm(24.0),
        area_height_min=inch2mm(3.0),
        area_height_max=inch2mm(40.0),
        tool_depth_min=0,
        tool_depth_max=10,
        tool_speed_max=3000.0,
    ),
```

**Comment by mossblaser** [comment]
Hey there,

Thank you both for adding these device details! You should find them now released in py_silhouette 0.6.0! Please let me know if you have any issues!

I also hope you're OK with the [comments crediting your contributions](https://github.com/mossblaser/py_silhouette/blob/master/py_silhouette/device.py#L259-L286).

Thanks again and happy plotting!

**Comment by arthursw** [comment]
Wonderful ! Thanks ;)
